storage/mysql/log_storage.go

```diff
@@ -323,11 +320,14 @@ func (t *logTreeTX) QueueLeaves(leaves []*trillian.LogLeaf, queueTimestamp time.
 		_, err := t.tx.Exec(insertSQL, t.treeID, leaf.LeafIdentityHash, leaf.LeafValue, leaf.ExtraData)
 		if err != nil {
 			if strings.Contains(err.Error(), "Duplicate entry") {
-				return grpc.Errorf(codes.AlreadyExists, "Leaf already exists for IdentityHash: %x. err: %v",
-					leaf.LeafIdentityHash, err)
+				return storage.Error{
+					ErrType: storage.DuplicateLeaf,
+					Cause:   err,
+					Detail:  fmt.Sprintf("IdentityHash: %x", leaf.LeafIdentityHash),
+				}
```

Comment:
I was trying to make it non intrusive so it only intervened if we know it's an error that needs special handling. Could go through and wrap everything in an error of UNKNOWN but type wasn't convinced that this is useful.